File: src/Capture.java
public class Capture implements Event {
    private final Position removedPosition;
    private Piece removedPiece;
    public Capture (Position removedPosition) {
        this.removedPosition = removedPosition;
    }

    @Override
    public void doToBoard(Board board) {
        removedPiece = board.atPosition(removedPosition);
        board.replaceAtPosition(removedPosition, new Empty());
    }

    @Override
    public void undoToBoard(Board board) {
        board.replaceAtPosition(removedPosition, removedPiece);
    }
}

File: src/Capture.kt
class Capture(private val removedPosition: Position) : Event {
    private var removedPiece: Piece? = null
    override fun doToBoard(board: Board) {
        removedPiece = board.atPosition(removedPosition)
        board.replaceAtPosition(removedPosition, Empty())
    }

    override fun undoToBoard(board: Board) {
        board.replaceAtPosition(removedPosition, removedPiece!!)
    }
}